### DISCUSSION TITLE
Test Double For User and AuthDataResult classes

### DISCUSSION THREAD

**Discussion by kerimsari** [firebase/firebase-ios-sdk#12073] [Q&A]
Hello everybody,

We are using **firebase authentication** sdk for swift in our project.  I've succeeded to mock FIRAuth but i wasn't able to write spy classes for User and AuthDataResult of firebase auth sdk in swift.

Please suggest....

Kerim

**Comment by paulb777** [comment]
Hi @kerimsari We've been working on moving the Auth implementation to Swift and you may find some pointers in the branch at https://github.com/firebase/firebase-ios-sdk/tree/auth-swift

**Reply by kerimsari** [reply]
Thanks Paul.

Do you know where exactly are fake classes for those 2 types AuthDataResult and User?

**Reply by paulb777** [reply]
Might not have what you're looking for, since most of the testing there is unit testing and fakes are usually done at a lower-level(like networking calls).

**Reply by kerimsari** [reply]
So how can we isolate firebase ios sdk classes and test our code? What do you suggest?

Thank you

**Reply by ncooke3** [reply]
Hi @kerimsari, could you share some of the challenges you are facing when trying to write spy classes for `User` and `AuthDataResult`?

**Reply by kerimsari** [reply]
Hi @ncooke3 the basic problem that i face is that i can not create a fake class derived from the same interfaces as AuthDataResult and User in Swift

First problem is that they are not initializable. 
Second problem is that there is a user inside AuthDataResult but it is encapsulated and not accesible.

So even i define 2 separate protocols and extend those objective-c interfaces (AuthDataResult and User) with these 2 protocols and derive my spies from those same protocols, i get invalid casting issues as the parent classes are different. 

So it would be great if you can achieve to write a fake class that i will be able to pass as a stub to the spy class belove

protocol FirebaseAuthenticationProtocol {
    var currentUser: FirebaseAuth.User? { get }
    
    func createUser(withEmail email: String, password: String, completion: ((AuthDataResult?, Error?) -> Void)?)
    func fetchSignInMethods(forEmail email: String, completion: (([String]?, Error?) -> Void)?)
    func signIn(withEmail email: String, password: String, completion: ((AuthDataResult?, Error?) -> Void)?)
    func signOut() throws
    func sendPasswordReset(withEmail email: String, completion: ((Error?) -> Void)?)
        
    func addIDTokenDidChangeListener(_ listener: @escaping (Auth, FirebaseAuth.User?) -> Void) -> IDTokenDidChangeListenerHandle
}

import FirebaseAuth

class SpyFirebaseAuth: FirebaseAuthenticationProtocol {
    var currentUser: FirebaseAuth.User?
    
    var sendResetPasswordCalled: Bool = false
    
    var fetchSignInMethodsError: Error? = nil
    var signInMethods: [String]? = nil
    
    private(set) var fetchSignInMethodsCalled: Bool = false
    
    private(set) var createUserCalled: Bool = false
       
    var createUserError: Error? = nil
    var createUserResult: AuthDataResult? = nil
    
    func signIn(withEmail email: String, password: String, completion: ((AuthDataResult?, Error?) -> Void)? = nil) {}
    
    func createUser(withEmail email: String, password: String, completion: ((AuthDataResult?, Error?) -> Void)? = nil) {
        createUserCalled = true
        completion?(createUserResult, createUserError)
    }
    
    func sendPasswordReset(withEmail email: String, completion: ((Error?) -> Void)? = nil) {
        sendResetPasswordCalled = true
        completion?(nil)
    }
    
    func fetchSignInMethods(forEmail email: String, completion: (([String]?, Error?) -> Void)? = nil) {
        fetchSignInMethodsCalled = true
        completion?(signInMethods, fetchSignInMethodsError)
    }
    
    func signOut() throws {}
    
    func addIDTokenDidChangeListener(_ listener: @escaping (Auth, FirebaseAuth.User?) -> Void) -> IDTokenDidChangeListenerHandle {
        return Auth.auth().addStateDidChangeListener(listener)
    }
}

Thank you for your supprt!
Kerim
